lib/opal/cli_runners/chrome_cdp_interface.rb

```diff
@@ -3,143 +3,185 @@
 # This script I converted into Opal, so that I don't have to write
 # buffer handling again. We have gets, Node has nothing close to it,
 # even async.
+# For CDP see docs/cdp_common.(md|json)
 
 require 'opal/platform'
 
 %x{
 var CDP = require("chrome-remote-interface");
 var fs = require("fs");
+var http = require("http");
 
 var dir = #{ARGV.last}
+var offset; // port offset for http server, depending on number of targets
 
 var options = {
   host: #{ENV['CHROME_HOST'] || 'localhost'},
   port: #{ENV['CHROME_PORT'] || 9222}
 };
 
+function not_found(res) {
+  res.writeHead(400, { "Content-Type": "text/plain" });
+  res.end("NOT FOUND");
+}
+
+var server = http.createServer(function(req, res) {
+  if (req.method === "GET") {
+    var path = dir + '/' + req.url.slice(1);
+    if (path.includes('..') || !fs.existsSync(path)) {
+      not_found(res);
+    } else {
+      var content_type;
+      if (path.endsWith(".html")) {
+        content_type = "text/html"
+      } else if (path.endsWith(".map")) {
+        content_type = "application/json"
+      } else {
+        content_type = "application/javascript"
+      }
+      res.writeHead(200, { "Content-Type": content_type });
+      res.end(fs.readFileSync(path));
+    }
+  } else if (req.method === "POST") {
+    if (req.url === "/File.write") {
```

Comment:
We need POST so that the browser can write bnechmark results to a file.
benchmark_driver creates a ruby file with the benchmark, thats executed and writes the times to a file, using File#write, which is then read by benchmark_driver to get the benchmark result.

Reply:
Please check [286dc35](https://github.com/opal/opal/pull/2461/commits/286dc350b7e0c5065c54203d76fbdbdc689e4690)
But well, CDP is locally open anyway, so you may use that, to get the secret ...

Anyway, that scheme doesn't seem to work with Firefox ...

Still investigating ...

Reply:
You could use CDP to overwrite any file using the screenshot, pdf or save page functions

Reply:
No, i am wrong, they changed that, screenshot, pdf return a io stream, but https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-setDownloadBehavior alloes for any path